### PR TITLE
fix: Add missing multiprocessing/resource-tracker type hints

### PR DESCRIPTION
**Pull request by kkirsche**
X-Ref: https://github.com/python/typeshed/issues/6799

This adds best-effort types to `multiprocessing.resource_tracker` to begin to address issue https://github.com/python/typeshed/issues/6799

The source file can be viewed here:
https://github.com/python/cpython/blob/main/Lib/multiprocessing/resource_tracker.py

### CODE REVIEW

**Comment by kkirsche** [comment]
It sounds like this is Python 3.8+ only from this error:

```
error: multiprocessing.resource_tracker failed to import, ModuleNotFoundError: No module named 'multiprocessing.resource_tracker'
Stub: at line 1
MypyFile:1(
  stdlib/multiprocessing/resource_tracker.pyi)
Runtime:
MISSING
```

Before making changes though, would someone be able to confirm the correct way to address this? I want to avoid unnecessary emails and whatnot as I try to work it out myself

**Comment by JelleZijlstra** [comment]
You may need to add it to the VERSIONS file if it's new in 3.8.

**Comment by kkirsche** [comment]
Seems to be the case:
https://github.com/python/cpython/blob/3.8/Lib/multiprocessing/resource_tracker.py
vs. 404 for
https://github.com/python/cpython/blob/3.7/Lib/multiprocessing/resource_tracker.py

Fixed in d5294ead

File: stdlib/multiprocessing/resource_tracker.pyi [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+from threading import Lock
+from typing import Any
+
+__all__ = ["ensure_running", "register", "unregister"]
+
+_HAVE_SIGMASK: bool = ...
+_IGNORED_SIGNALS: tuple[Signals, Signals] = ...
+
+_CLEANUP_FUNCS: dict[str, Callable[[Any], Any]] = ...
+
```
Comment (AlexWaygood) [inline stdlib/multiprocessing/resource_tracker.pyi:13]: Same as https://github.com/python/typeshed/pull/8406#discussion_r931170447

```suggestion
```

File: stdlib/multiprocessing/resource_tracker.pyi [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+class ResourceTracker:
+    _lock: Lock
+    _fd: int | None
+    _pid: int | None
+
```
Comment (AlexWaygood) [inline stdlib/multiprocessing/resource_tracker.pyi:18]:
```suggestion
```

File: stdlib/multiprocessing/resource_tracker.pyi [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    def __init__(self) -> None: ...
+    def _stop(self) -> None: ...
```
Comment (AlexWaygood) [inline stdlib/multiprocessing/resource_tracker.pyi:20]:
```suggestion
```

File: stdlib/multiprocessing/resource_tracker.pyi [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    def getfd(self) -> int | None: ...
+    def ensure_running(self) -> None: ...
+    def _check_alive(self) -> bool: ...
```
Comment (AlexWaygood) [inline stdlib/multiprocessing/resource_tracker.pyi:23]:
```suggestion
```

File: stdlib/multiprocessing/resource_tracker.pyi [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    def register(self, name: Sized, rtype: object) -> None: ...
+    def unregister(self, name: Sized, rtype: object) -> None: ...
+    def _send(self, cmd: object, name: Sized, rtype: object) -> None: ...
```
Comment (AlexWaygood) [inline stdlib/multiprocessing/resource_tracker.pyi:26]:
```suggestion
```

File: stdlib/multiprocessing/resource_tracker.pyi [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+_resource_tracker: ResourceTracker = ...
+ensure_running: Callable[[], None] = ...
+register: Callable[[Sized, object], None] = ...
+unregister: Callable[[Sized, object], None] = ...
+getfd: Callable[[], int | None] = ...
```
Comment (AlexWaygood) [inline stdlib/multiprocessing/resource_tracker.pyi:32]: You had the right idea in #8406 for how to tackle bound methods :)

```suggestion
ensure_running = _resource_tracker.ensure_running
register = _resource_tracker.register
unregister = _resource_tracker.unregister
getfd = _resource_tracker.getfd
```

**Review by AlexWaygood — CHANGES_REQUESTED** [review]

File: stdlib/multiprocessing/resource_tracker.pyi [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+from collections.abc import Sized
+
+__all__ = ["ensure_running", "register", "unregister"]
+
+class ResourceTracker:
+    def __init__(self) -> None: ...
+    def getfd(self) -> int | None: ...
+    def ensure_running(self) -> None: ...
+    def register(self, name: Sized, rtype: object) -> None: ...
+    def unregister(self, name: Sized, rtype: object) -> None: ...
```
Comment (AlexWaygood) [inline stdlib/multiprocessing/resource_tracker.pyi:11]: It's true that these methods can be passed literally any object without raising an exception, so your instinct here to annotate these parameters with `object` was good. However, I feel like it's unlikely that you're actually _meant_ to be passing literally any object to these methods. I think it's more likely that you're meant to be passing some kind of "resource" to these methods (the "shared memory segments, semaphores, etc." referenced in the block comment at the top of the module). I don't know enough about `multiprocessing` to know what the _true_ type is, so I'd rather we went for `_typeshed.Incomplete` for now, which serves as an explicit marker of "we're not quite sure what this type should be".

```suggestion
    def register(self, name: Sized, rtype: Incomplete) -> None: ...
    def unregister(self, name: Sized, rtype: Incomplete) -> None: ...
```

Comment (kkirsche) [reply stdlib/multiprocessing/resource_tracker.pyi]: That makes sense, updated :) thanks for explaining that 👍

**Review by AlexWaygood — APPROVED** [review]

**Pull request merged**